### DISCUSSION TITLE
Rgroupdecomposition questions

### DISCUSSION THREAD

**Discussion by orgw** [rdkit/rdkit#4749] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/64889801/143972915-61169b8b-031a-4a20-a12b-d09771f265d3.png)

Hi i;m knew to rdkit and i'd like to ask some questions

1) first question is, after getting the core, i did r group decomposition but as you can see in R1, the aromatic rings are gone.

The solution i thought would be is to change the core...but I don't know how. I tried to use Chem.MolToSmiles(scaffold) to get the smiles format, change it and then insert but i get valence errors..

'CC(C)NC1:N:C:N:C2:C:C:C:C:C:1:2'

to

'C(C)NC1:N:C:N:C2:C:C:C:C:C:1:2'

gets error RDKit ERROR: [11:10:48] Explicit valence for atom # 7 C, 5, is greater than permitted
[11:10:48] Explicit valence for atom # 7 C, 5, is greater than permitted

smarts pattern is 

'[#6&R]1:&@[#6&R]:&@[#6&R]2:&@[#6&R](:&@[#7&R]:&@[#6&R]:&@[#7&R]:&@[#6&R]:&@2:&@[#6&R]:&@[#6&R]:&@1)-&!@[#7&!R]-&!@[#6&!R](-&!@[#6&!R])-&!@[#6&R]'

2) second problem is that i get unwanted R groups such as R3, which is all hydrogen. 

How do i specify?? Thanks

**Comment by bp-kelley** [comment, marked as answer]
The issue is that the core has an atom in the aromatic ring and the RGroup is every atom in the ring but that one.  The rendering looks wrong because RDKit looses the fact that it is a ring, but  the dashed lines indicate aromaticity.

If you remove that single atom from the core, you'll solve your issue.

You can specify the core as a smarts, which will work with the example structures you gave, smarts don't have to be real molecules, so they won't throw valence errors.  The RGroupDecomposition tool can use either smiles or smarts inputs.

```
core = Chem.MolFromSmarts('C(C)NC1:N:C:N:C2:C:C:C:C:C:1:2')
```

I'm curious where the original core came from?

**Reply by orgw** [reply]
Awesome. Thanks

**Reply by orgw** [reply]
I got the core by MCS

**Comment by TAEUKHAM** [comment]
For question 2, you could change rGroupDecomposition options. If you want to ignore rgroups only formed in hydrogen, you might try this.
```
from rdkit.Chem import rdRGroupDecomposition
params = rdRGroupDecomposition.RGroupDecompositionParameters()
params.removeAllHydrogenRGroups=True
params.removeAllHydrogenRGroupsAndLabels =True
params.removeHydrogensPostMatch = True
```

**Reply by orgw** [reply]
Thanks a lot!

**Reply by orgw** [reply]
![image](https://user-images.githubusercontent.com/64889801/143985319-889eceb3-e148-49df-84c8-3c3968fe7b12.png)

I tried your solution, but doesn't seem to work. maybe it's because of R3 label?

**Reply by TAEUKHAM** [reply]
Hmmm... What is the "R3" value when you tried this option? If the value is None, you could add the 
```
params.onlyMatchAtRGroups = False
```
to delete the none term residual.

If this also doesn't work, I think you should remove H(by using RemoveHs method) of the compound before you decompose it.

**Reply by TAEUKHAM** [reply]
Oh, and I wonder why did you set the allowNonTerminalRGroups as True. Try setting it to false since you don't want the residuals to be as None value.

**Reply by orgw** [reply]
R3 is the H group!
